src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/BuildHostProcessManager.cs

```diff
@@ -0,0 +1,269 @@
+        using (var fileStream = new FileStream(projectFilePath, FileMode.Open, FileAccess.Read))
+        using (var xmlReader = XmlReader.Create(fileStream, s_xmlSettings))
+            document = XDocument.Load(xmlReader);
+
+        // If we don't have a root, doesn't really matter which. This project is just malformed.
+        if (document.Root == null)
+            return BuildHostProcessKind.NetCore;
+
+        // Look for SDK attribute on the root
+        if (document.Root.Attribute("Sdk") != null)
+            return BuildHostProcessKind.NetCore;
+
+        // Look for <Import Sdk=... />
+        if (document.Root.Elements("Import").Attributes("Sdk").Any())
+            return BuildHostProcessKind.NetCore;
+
+        // Look for <Sdk ... />
+        if (document.Root.Elements("Sdk").Any())
+            return BuildHostProcessKind.NetCore;
+
+        // Looking for PropertyGroups that contain TargetFramework or TargetFrameworks nodes
+        var propertyGroups = document.Descendants("PropertyGroup");
+        if (propertyGroups.Elements("TargetFramework").Any() || propertyGroups.Elements("TargetFrameworks").Any())
+            return BuildHostProcessKind.NetCore;
+
+        // Nothing that indicates it's an SDK-style project, so use our .NET framework host
+        return BuildHostProcessKind.NetFramework;
+    }
+
+    private enum BuildHostProcessKind
+    {
+        NetCore,
+        NetFramework
+    }
+
+    private sealed class BuildHostProcess : IAsyncDisposable
+    {
+        private readonly Process _process;
+        private readonly JsonRpc _jsonRpc;
+
+        private int _disposed = 0;
+
+        public BuildHostProcess(Process process, ILoggerFactory? loggerFactory)
+        {
+            LoggerForProcessMessages = loggerFactory?.CreateLogger($"BuildHost PID {process.Id}");
+
+            _process = process;
+
+            _process.EnableRaisingEvents = true;
+            _process.Exited += Process_Exited;
+
+            _process.ErrorDataReceived += Process_ErrorDataReceived;
+
+            var messageHandler = new HeaderDelimitedMessageHandler(sendingStream: _process.StandardInput.BaseStream, receivingStream: _process.StandardOutput.BaseStream, new JsonMessageFormatter());
+
+            _jsonRpc = new JsonRpc(messageHandler);
+            _jsonRpc.StartListening();
+            BuildHost = _jsonRpc.Attach<IBuildHost>();
+
+            // Call this last so our type is fully constructed before we start firing events
+            _process.BeginErrorReadLine();
+        }
+
+        private void Process_Exited(object? sender, EventArgs e)
+        {
+            Disconnected?.Invoke(this, EventArgs.Empty);
+        }
+
+        private void Process_ErrorDataReceived(object sender, DataReceivedEventArgs e)
+        {
+            if (e.Data is not null)
+                LoggerForProcessMessages?.LogTrace($"Message from Process: {e.Data}");
```

Comment:
Something to consider for later - similar to LSP trace logging, sometimes the json itself is useful to log at the most verbose levels.  In case there are weird serialization issues.